sys/linux/dev_i915.txt

```diff
@@ -0,0 +1,360 @@
+ioctl$DRM_IOCTL_I915_OVERLAY_PUT_IMAGE(fd fd_i915, cmd const[DRM_IOCTL_I915_OVERLAY_PUT_IMAGE], arg ptr[in, drm_intel_overlay_put_image])
+ioctl$DRM_IOCTL_I915_OVERLAY_ATTRS(fd fd_i915, cmd const[DRM_IOCTL_I915_OVERLAY_ATTRS], arg ptr[in, drm_intel_overlay_attrs])
+ioctl$DRM_IOCTL_I915_SET_SPRITE_COLORKEY(fd fd_i915, cmd const[DRM_IOCTL_I915_SET_SPRITE_COLORKEY], arg ptr[in, drm_intel_sprite_colorkey])
+ioctl$DRM_IOCTL_I915_GEM_WAIT(fd fd_i915, cmd const[DRM_IOCTL_I915_GEM_WAIT], arg ptr[inout, drm_i915_gem_wait])
+ioctl$DRM_IOCTL_I915_GEM_CONTEXT_CREATE(fd fd_i915, cmd const[DRM_IOCTL_I915_GEM_CONTEXT_CREATE], arg ptr[in, drm_i915_gem_context_create])
+ioctl$DRM_IOCTL_I915_GEM_CONTEXT_DESTROY(fd fd_i915, cmd const[DRM_IOCTL_I915_GEM_CONTEXT_DESTROY], arg ptr[in, drm_i915_gem_context_destroy])
+ioctl$DRM_IOCTL_I915_REG_READ(fd fd_i915, cmd const[DRM_IOCTL_I915_REG_READ], arg ptr[inout, drm_i915_reg_read])
+ioctl$DRM_IOCTL_I915_GET_RESET_STATS(fd fd_i915, cmd const[DRM_IOCTL_I915_GET_RESET_STATS], arg ptr[inout, drm_i915_reset_stats])
+ioctl$DRM_IOCTL_I915_GEM_CONTEXT_GETPARAM(fd fd_i915, cmd const[DRM_IOCTL_I915_GEM_CONTEXT_GETPARAM], arg ptr[in, drm_i915_gem_context_param])
+ioctl$DRM_IOCTL_I915_GEM_CONTEXT_SETPARAM(fd fd_i915, cmd const[DRM_IOCTL_I915_GEM_CONTEXT_SETPARAM], arg ptr[inout, drm_i915_gem_context_param])
+ioctl$DRM_IOCTL_I915_PERF_OPEN(fd fd_i915, cmd const[DRM_IOCTL_I915_PERF_OPEN], arg ptr[in, drm_i915_perf_open_param])
+ioctl$DRM_IOCTL_I915_PERF_ADD_CONFIG(fd fd_i915, cmd const[DRM_IOCTL_I915_PERF_ADD_CONFIG], arg ptr[in, drm_i915_perf_oa_config])
+ioctl$DRM_IOCTL_I915_GEM_USERPTR(fd fd_i915, cmd const[DRM_IOCTL_I915_GEM_USERPTR], arg ptr[inout, drm_i915_gem_userptr])
+ioctl$DRM_IOCTL_I915_PERF_REMOVE_CONFIG(fd fd_i915, cmd const[DRM_IOCTL_I915_PERF_REMOVE_CONFIG], arg ptr[out, int64])
+ioctl$DRM_IOCTL_I915_QUERY(fd fd_i915, cmd const[DRM_IOCTL_I915_QUERY], arg ptr[inout, drm_i915_query])
+ioctl$DRM_IOCTL_I915_GEM_VM_CREATE(fd fd_i915, cmd const[DRM_IOCTL_I915_GEM_VM_CREATE], arg ptr[inout, drm_i915_gem_vm_control])
+ioctl$DRM_IOCTL_I915_GEM_VM_DESTROY(fd fd_i915, cmd const[DRM_IOCTL_I915_GEM_VM_DESTROY], arg ptr[out, drm_i915_gem_vm_control])
+
+mmap$DRM_I915(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_i915, offset fileoff)
+_ = __NR_mmap2
+
+type i915_gem_ctx_id int32
+type i915_handle int32
+
+drm_i915_gem_busy {
+	handle	i915_handle
+	busy	int32
+}
+
+drm_i915_gem_caching {
+	handle	i915_handle
+	caching	int32
+}
+
+drm_i915_gem_context_create {
+	ctx_id	i915_gem_ctx_id
+	pad	const[0, int32]
+}
+
+drm_i915_gem_context_destroy {
+	ctx_id	i915_gem_ctx_id
+	pad	const[0, int32]
+}
+
+drm_i915_gem_context_param {
+	ctx_id	i915_gem_ctx_id
+	size	int32
+	param	flags[i915_gem_param_flags, int64]
+	value	int64
+}
+
+drm_i915_gem_create {
+	size	int64
+	handle	i915_handle
+	pad	const[0, int32]
+}
+
+drm_i915_gem_execbuffer {
+	buffers_ptr		ptr[in, array[drm_i915_gem_exec_object]]
```

Comment:
This should be `ptr64`, it's 8 bytes regardless of bitness.
It seems to be a common problem here:
```
bad-struct-size: drm_i915_gem_execbuffer: syz=32 kernel=40 [386]
bad-field-size: drm_i915_gem_execbuffer.buffers_ptr: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_execbuffer.buffer_count: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_execbuffer.batch_start_offset: syz=8 kernel=12 [386]
bad-field-offset: drm_i915_gem_execbuffer.batch_len: syz=12 kernel=16 [386]
bad-field-offset: drm_i915_gem_execbuffer.DR1: syz=16 kernel=20 [386]
bad-field-offset: drm_i915_gem_execbuffer.DR4: syz=20 kernel=24 [386]
bad-field-offset: drm_i915_gem_execbuffer.num_cliprects: syz=24 kernel=28 [386]
bad-field-offset: drm_i915_gem_execbuffer.cliprects_ptr: syz=28 kernel=32 [386]
bad-field-size: drm_i915_gem_execbuffer.cliprects_ptr: syz=4 kernel=8 [386]
bad-struct-size: drm_i915_gem_exec_object: syz=28 kernel=32 [386]
bad-field-size: drm_i915_gem_exec_object.relocs_ptr: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_exec_object.alignment: syz=12 kernel=16 [386]
bad-field-offset: drm_i915_gem_exec_object.offset: syz=20 kernel=24 [386]
bad-struct-size: drm_i915_gem_execbuffer2: syz=56 kernel=64 [386]
bad-field-size: drm_i915_gem_execbuffer2.buffers_ptr: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_execbuffer2.buffer_count: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_execbuffer2.batch_start_offset: syz=8 kernel=12 [386]
bad-field-offset: drm_i915_gem_execbuffer2.batch_len: syz=12 kernel=16 [386]
bad-field-offset: drm_i915_gem_execbuffer2.DR1: syz=16 kernel=20 [386]
bad-field-offset: drm_i915_gem_execbuffer2.DR4: syz=20 kernel=24 [386]
bad-field-offset: drm_i915_gem_execbuffer2.num_cliprects: syz=24 kernel=28 [386]
bad-field-offset: drm_i915_gem_execbuffer2.cliprects_ptr: syz=28 kernel=32 [386]
bad-field-size: drm_i915_gem_execbuffer2.cliprects_ptr: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_execbuffer2.flags: syz=32 kernel=40 [386]
bad-field-offset: drm_i915_gem_execbuffer2.rsvd1: syz=40 kernel=48 [386]
bad-field-offset: drm_i915_gem_execbuffer2.rsvd2: syz=48 kernel=56 [386]
bad-struct-size: drm_i915_gem_exec_object2: syz=52 kernel=56 [386]
bad-field-size: drm_i915_gem_exec_object2.relocs_ptr: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_exec_object2.alignment: syz=12 kernel=16 [386]
bad-field-offset: drm_i915_gem_exec_object2.offset: syz=20 kernel=24 [386]
bad-field-offset: drm_i915_gem_exec_object2.flags: syz=28 kernel=32 [386]
bad-field-offset: drm_i915_gem_exec_object2.rsvd1/: syz=36 kernel=40 [386]
bad-field-offset: drm_i915_gem_exec_object2.rsvd2: syz=44 kernel=48 [386]
bad-struct-size: drm_i915_gem_mmap: syz=36 kernel=40 [386]
bad-field-size: drm_i915_gem_mmap.addr_ptr: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_mmap.flags: syz=28 kernel=32 [386]
bad-struct-size: drm_i915_gem_pread: syz=28 kernel=32 [386]
bad-field-size: drm_i915_gem_pread.data_ptr: syz=4 kernel=8 [386]
bad-struct-size: drm_i915_gem_pwrite: syz=28 kernel=32 [386]
bad-field-size: drm_i915_gem_pwrite.data_ptr: syz=4 kernel=8 [386]
bad-struct-size: drm_i915_gem_userptr: syz=20 kernel=24 [386]
bad-field-size: drm_i915_gem_userptr.user_ptr: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_userptr.user_size: syz=4 kernel=8 [386]
bad-field-offset: drm_i915_gem_userptr.flags: syz=12 kernel=16 [386]
bad-field-offset: drm_i915_gem_userptr.handle: syz=16 kernel=20 [386]
bad-struct-size: drm_i915_perf_open_param: syz=12 kernel=16 [386]
bad-field-size: drm_i915_perf_open_param.properties_ptr: syz=4 kernel=8 [386]
bad-struct-size: drm_i915_query: syz=12 kernel=16 [386]
bad-field-size: drm_i915_query.items_ptr: syz=4 kernel=8 [386]
bad-struct-size: drm_i915_query_item: syz=20 kernel=24 [386]
bad-field-size: drm_i915_query_item.data_ptr: syz=4 kernel=8 [386]
```

Reply:
Thanks! I've fixed ptr usages to ptr64.

I re-tried using syz-check(`syz-check -obj-amd64 ${KSRC}/vmlinux`) but was not able to get the above output, I'm guessing I have not set it up correctly. The kernel is built with debug symbols and I'm able to see the struct information using pahole( eg: `pahole ./vmlinux -C drm_i915_gem_busy`), but syz-check does not produce any warn information. Based on the info at [link](https://pkg.go.dev/github.com/google/syzkaller@v0.0.0-20200622131031-1afe15356a4b/tools/syz-check?tab=doc), I set `CFLAGS_MODULE` and `CFLAGS_KERNEL` to `-ggdb -fno-eliminate-unused-debug-types -fno-eliminate-unused-debug-symbols`

Reply:
These were warning with 32-bit i386 build. For 64-bit builds ptr and ptr64 are the same.

Reply:
syz-check accepts multiple object files are the same time, I run it as:
```
syz-check -obj-386 ~/src/linux_386/vmlinux \
	-obj-amd64 ~/src/linux_amd64/vmlinux \
	-obj-arm ~/src/linux_arm/vmlinux \
	-obj-arm64 ~/src/linux_arm64/vmlinux
```
(you can pass any subset of these)
This is very inconvenient to run at the moment, so I usually don't ask contributors to do this manually :)